mycodo/utils/method.py

```diff
@@ -236,65 +285,165 @@ def calculate_method_setpoint(method_id, table, this_controller, Method, MethodD
+                        "[Method] Percent of row: {per}".format(
+                            per=percent_row))
+                    self.logger.debug(
+                        "[Method] New Setpoint: {sp}".format(
+                            sp=new_setpoint))
                 return new_setpoint, False
             previous_total_sec = total_sec
 
-        if this_controller.method_start_time:
-            if method_restart:
-                if end_time and now > end_time:
-                    ended = True
-                else:
-                    # Method has been instructed to restart
-                    with session_scope(MYCODO_DB_PATH) as db_session:
-                        mod_method = db_session.query(table)
-                        mod_method = mod_method.filter(
-                            table.unique_id == this_controller.unique_id).first()
-                        mod_method.method_start_time = datetime.datetime.now()
-                        db_session.commit()
-                        return previous_end, False
+        return None, True
+
+    def should_restart(self, now, method_start_time=None):
+        for each_method in self.method_data_all:
+            # If duration_sec is 0, method has instruction to restart
+            if each_method.duration_sec == 0:
+                return True
+
+        return False
+
+    def get_plot(self, max_points_x=None):
+        result = []
+        first_entry = True
+        start_duration = 0
+        for each_method in self.method_data_all:
+            if each_method.setpoint_end is None:
+                setpoint_end = each_method.setpoint_start
+            else:
+                setpoint_end = each_method.setpoint_end
+
+            if each_method.duration_sec == 0:
+                pass  # Method line is repeat command, don't add to method_list
+            elif first_entry:
+                result.append([0, each_method.setpoint_start])
+                result.append([each_method.duration_sec, setpoint_end])
+                start_duration += each_method.duration_sec
+                first_entry = False
             else:
-                ended = True
+                end_duration = start_duration + each_method.duration_sec
+
+                result.append(
+                    [start_duration, each_method.setpoint_start])
+                result.append(
+                    [end_duration, setpoint_end])
+
+                start_duration += each_method.duration_sec
+
+        return result
+
+
+def method_by_type(method_data, method_type, logger=None):
+    """
+    Looks up a method class suitable to the given method_type and initializes it with the given method_data
+    """
+
+    method_class = globals().get(method_type+"Method")
+    if not method_class or not issubclass(method_class, AbstractMethod):
+        logger.error("Method {} is unknown.".format(method_type))
+        method_class = AbstractMethod
+
+    return method_class(method_data, method_type, logger)
```

Comment:
This looks up the responsible class for a method_type dynamically. It saves us from if-then-else'ing though all known types. Add a class "xyzMethod" and it will be found for method_type "xyz".
**Note: Other places throughout the front-end code still do implement if-then-else lookup for the type. You still have to announce your new methed in these places.**